### PR TITLE
[benchmark] Use local narwhal channel in benchmark client

### PR DESCRIPTION
**Pull request by arun-koshy**
Benchmark channel will be started in process with the primary & worker so that a channel can be used to send transactions to the worker instead of going over the network. This will allow us to achieve higher throughput numbers.

Additionally 
- Increase socket buffer size for hosts used in benchmark
- Fix error reporting for connection failures
- Update narwhal grafana dashboard

### CODE REVIEW

File: crates/sui-aws-orchestrator/src/protocol/narwhal.rs [hunk 1]
```diff
@@ -109,14 +109,27 @@ impl ProtocolCommands<NarwhalBenchmarkType> for NarwhalProtocol {
...
-        _parameters: &BenchmarkParameters<NarwhalBenchmarkType>,
+        parameters: &BenchmarkParameters<NarwhalBenchmarkType>,
     ) -> Vec<(Instance, String)>
     where
         I: IntoIterator<Item = Instance>,
     {
         let working_dir = self.working_dir.clone();
+        let hosts: Vec<_> = instances.into_iter().collect();
+        // 2 ports used per authority so add 2 * num authorities to base port
+        let mut worker_base_port = BASE_PORT + (2 * hosts.len());
```
Comment (mwtian) [inline crates/sui-aws-orchestrator/src/protocol/narwhal.rs:120]: Why ports have to be different for each authority? Is it for supporting running on a single host?

File: narwhal/node/src/benchmark_client.rs [hunk 2]
```diff
@@ -53,6 +59,68 @@ struct App {
...
+        let client = client.as_ref().unwrap().load();
+        client
+            .submit_transaction(transaction)
+            .await
+            .map_err(|e| eyre::Report::msg(format!("Failed to submit to consensus: {:?}", e)))?;
+        Ok(())
+    }
+}
+
+#[allow(dead_code)]
```
Comment (mwtian) [inline narwhal/node/src/benchmark_client.rs:123]: Why is this needed?

File: narwhal/node/src/benchmark_client.rs [hunk 3]
```diff
@@ -53,6 +59,68 @@ struct App {
     client_metric_port: u16,
 }
 
+/// A Narwhal client that instantiates LocalNarwhalClient lazily. (taken from consensus adapter)
+pub struct LazyNarwhalClient {
```
Comment (mwtian) [inline narwhal/node/src/benchmark_client.rs:63]: I hope this code can be reused in some way. e.g. maybe we can move the code to worker instead.

File: narwhal/node/src/benchmark_client.rs [hunk 4]
```diff
@@ -142,21 +217,29 @@ impl Client {
         // TODO: figure out how to scale the client without needing to scale tasks
         // Current results are showing about 10 tx/s per task.
         let num_parallel_tasks = self.rate.min(25000);
```
Comment (mwtian) [inline narwhal/node/src/benchmark_client.rs:219]: Are there 25000 tasks across all nodes?

File: narwhal/node/src/benchmark_client.rs [hunk 5]
```diff
@@ -142,21 +217,29 @@ impl Client {
...
-            let mut client = TransactionsClient::connect(self.target.as_str().to_owned())
+            let task_rate = if i < remaining_transactions {
+                base_rate_per_task + 1
+            } else {
+                base_rate_per_task
+            };
+            let task_interval = Duration::from_millis(1000 / task_rate);
+
+            let local_client_clone = Arc::clone(&self.local_client);
+            let mut grpc_client = TransactionsClient::connect(self.target.as_str().to_owned())
```
Comment (mwtian) [inline narwhal/node/src/benchmark_client.rs:242]: Can we avoid using grpc_client at all?

**Review by mwtian — APPROVED** [review]
Great improvements!

File: crates/sui-core/src/authority_server.rs [hunk 6]
```diff
@@ -35,11 +35,12 @@ use tap::TapFallible;
 use tokio::task::JoinHandle;
 use tracing::{error_span, info, Instrument};
 
-use crate::consensus_adapter::{ConnectionMonitorStatusForTests, LazyNarwhalClient};
+use crate::consensus_adapter::ConnectionMonitorStatusForTests;
 use crate::{
     authority::AuthorityState,
     consensus_adapter::{ConsensusAdapter, ConsensusAdapterMetrics},
 };
+use narwhal_worker::LazyNarwhalClient;
```
Comment (mwtian) [inline crates/sui-core/src/authority_server.rs:43]: Maybe move this import to the section of non crate imports.

File: crates/sui-aws-orchestrator/src/protocol/narwhal.rs [hunk 7]
```diff
@@ -173,54 +196,17 @@ impl ProtocolCommands<NarwhalBenchmarkType> for NarwhalProtocol {
...
-                        Self::CLIENT_METRICS_PORT
-                    ),
-                ]
-                .join(" ");
-                let command = ["source $HOME/.cargo/env", &run].join(" && ");
-
-                (instance, command)
-            })
-            .collect()
+        vec![]
```
Comment (mwtian) [inline crates/sui-aws-orchestrator/src/protocol/narwhal.rs:205]: Leave a comment about why this is empty?

**Review by mwtian — APPROVED** [review]

File: narwhal/node/src/benchmark_client.rs [hunk 8]
```diff
@@ -51,8 +56,60 @@ struct App {
...
+        match s.to_lowercase().as_str() {
+            "local" => Ok(OperatingMode::Local),
+            "remote" => Ok(OperatingMode::Remote),
+            _ => Err("must be 'local' or 'remote'".to_string()),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait SubmitToConsensus: Sync + Send + 'static {
```
Comment (mwtian) [inline narwhal/node/src/benchmark_client.rs:85]: How is this trait definition and implementation used?

Comment (arun-koshy) [reply narwhal/node/src/benchmark_client.rs]: consensus_adapter.rs impl
```
async fn submit_to_consensus(
        &self,
        transaction: &ConsensusTransaction,
        _epoch_store: &Arc<AuthorityPerEpochStore>,
    ) -> SuiResult {}
```

It is just a custom trait so I can call `submit_to_consensus` from LazyNarwhalClient without having to pass in epoch store. But now that you ask I probably could have just created a method that took in LazyNarwhalClient as an arg. Will change it
